src/serial.rs

```diff
@@ -0,0 +1,484 @@
+    /// out one-by-one.
+    pub fn write<Ws>(words: Ws) -> Self
+    where
+        Ws: AsRef<[Word]>,
+    {
+        Transaction {
+            mode: words.as_ref().iter().cloned().map(Mode::Write).collect(),
+        }
+    }
+
+    /// Expecte a caller to flush the serial buffers
+    pub fn flush() -> Self {
+        Transaction {
+            mode: vec![Mode::Flush],
+        }
+    }
+}
+
+/// Mock serial device
+///
+/// The mock serial device can be loaded with expected transactions, then
+/// passed-on into a serial device user. If the expectations were not met
+/// in the specified order, the type causes a panic and describes what
+/// expectation wasn't met.
+///
+/// The type is clonable so that it may be shared with a serial
+/// device user. Under the hood, both cloned mocks will share
+/// the same state, allowing your handle to eventually call `done()`,
+/// if desired.
+pub struct Mock<'a, Word> {
+    /// The expected operations upon the mock
+    ///
+    /// It's in an arc to maintain shared state, and in a mutex
+    /// to make it thread safe. It's then wrapped in an `Option`
+    /// so that we can take it in the call to `done()`.
+    expected_modes: Arc<Mutex<Option<VecDeque<Mode<Word>>>>>,
+    /// We'll note that this phantom type and lifetime is actually
+    /// unnecessary. As noted above, we try to keep a consistent
+    /// public API with the existing spi and i2c modules. Given that
+    /// this lifetime is necessary for them, it is also necessary here.
+    ///
+    /// Of course, this could quickly be removed if the overall public
+    /// API changes.
+    marker: PhantomData<&'a Word>,
+}
+
+impl<'a, Word: Clone> Mock<'a, Word> {
+    /// Create a serial mock that will expect the provided transactions
+    pub fn new(transactions: &'a [Transaction<Word>]) -> Self {
+        let mut ser = Mock {
+            expected_modes: Arc::new(Mutex::new(None)),
+            marker: PhantomData,
+        };
+        ser.expect(transactions);
+        ser
+    }
+
+    /// Set expectations on the interface
+    ///
+    /// This is a list of transactions to be executed in order.
+    /// Note that setting this will overwrite any existing expectations
+    pub fn expect(&mut self, transactions: &'a [Transaction<Word>]) {
+        let mut lock = self
+            .expected_modes
+            .lock()
+            .expect("unable to lock serial mock in call to expect");
+        *lock = Some(
+            transactions
+                .iter()
+                .fold(VecDeque::new(), |mut modes, transaction| {
+                    modes.extend(transaction.mode.clone());
+                    modes
+                }),
+        );
+    }
+
+    /// Consumes the mock, asserting that all exepctations were met
+    pub fn done(self) {
+        // Note that we take self by value, unlike the generic versions that
+        // take by mut reference. It wouldn't make sense to call done() twice anyway :P
```

Comment:
Wouldn't re-using a mock instance be a valid use case?

For example:

```rust
mock.expect(...);
// do something with the mock
mock.done();
mock.expect(...);
// do some other things with the mock
mock.done();
```

For that to work, done would need to take a non-consuming reference to self.

Reply:
Thanks, I hadn't considered re-using the mock. It's definitely valid and useful after giving it a second look! I'll fix-up `done()`.